File: python/OWclient.py
#!/usr/bin/env python3
# author: Radoslav Grencik <xgrenc00@stud.fit.vutbr.cz>

################################################################################

import sys
import socket
import json
import time
import re

def Err(retval, msg):
    print("ERROR: {0}.".format(msg), file = sys.stderr)
    sys.exit(retval)
    
if len(sys.argv) != 3:
    Err(1, "bad args")

appid = sys.argv[1]
loc = sys.argv[2]

host = "api.openweathermap.org"
resource = "/data/2.5/weather?q=" + loc + "&APPID=" + appid +"&units=metric"

req = bytes("GET " + resource + " HTTP/1.1\r\n" + "Host: " + host + "\r\n\r\n", "utf-8") # HTTP request

soc = socket.socket(socket.AF_INET, socket.SOCK_STREAM) # adress (and protocol) family = AF_INET, socket type = SOCK_STREAM
try:
    soc.connect((host, 80)) # HTTP port = 80
except:
    Err(1, "can't connect")

soc.send(req)
result = soc.recv(4096) # size of buffer shoul be a small power of 2 = 4096
soc.close()

result = str(result, "utf-8")

check = result.split(" ")[1]
if check != "200":
    Err(check, check)

result = result.split("\r\n\r\n")[1]
jres = json.loads(result)

if ("deg" in result) == False:
    jres["wind"]["deg"] = "-"

sunrise = re.search("(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]", time.ctime(jres["sys"]["sunrise"])).group()
sunset = re.search("(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]:[0-5][0-9]", time.ctime(jres["sys"]["sunset"])).group()

print(
"""Weather in {0}, {1}
{2}
temp:       {3} °C
pressure:   {4} hPa
humidity:   {5} %
visibility: {6} m
wind-speed: {7} m/s
wind-deg:   {8}
cloudiness: {9} %
sunrise:    {10}
sunset:     {11}"""
.format(
jres["name"],
jres["sys"]["country"],
jres["weather"][0]["description"],
jres["main"]["temp"],
jres["main"]["pressure"],
jres["main"]["humidity"],
jres["visibility"],
jres["wind"]["speed"],
jres["wind"]["deg"],
jres["clouds"]["all"],
sunrise,
sunset
))

File: java/OWclient/src/owclient/OWclient.java
/**
 * @author Radoslav Grencik <xgrenc00@stud.fit.vutbr.cz>
 */
package owclient;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.net.Socket;
import java.util.Scanner;

/**
 *
 */
public class OWclient {

    private static String host, resource;
    private static final int PORT = 80;
    private static Socket s;

    public static void get() {
        try {
            InputStream input = s.getInputStream();
            OutputStream output = s.getOutputStream();
            try (Scanner sc = new Scanner(input)) {
                PrintWriter pw = new PrintWriter(output);

                String command = "GET " + resource + " HTTP/1.1\n" + "Host: " + host + "\n\n";
                pw.print(command);
                pw.flush();

                String l = "";
                for (int i = 0; i < 12; i++) {
                    l = sc.nextLine();
                }

                char[] arr = l.toCharArray();

                int i = 0, brac = 0;
                while (i < arr.length) {
                    if (arr[i] == '{') {
                        brac++;
                    }
                    if (arr[i] == '}') {
                        brac--;
                    }
                    if (arr[i] == '{' || arr[i] == '}' || arr[i] == '"' || arr[i] == '[' || arr[i] == ']') {
                        arr[i] = 0;
                    }

                    if (arr[i] == ',' && brac == 1) {
                        arr[i] = '\n';
                    }
                    System.out.printf("%c", arr[i]);
                    i++;
                }

                i = 0;
                int counter = 0;
                while (i < arr.length) {
                    if (arr[i] == '\n') counter++;
                    if (counter == 10) {
                        i = i + 9;
                        break;
                    }
                    i++;
                }

                char[] cityarr = new char[50];
                counter = 0;
                while (arr[i] != '\n') {
                    cityarr[counter] = arr[i];
                    System.out.printf("%c", cityarr[counter]);
                    i++;
                    counter++;
                }

                s.close();
            }
        } catch (IOException e) {
            System.err.println("Problem: " + e.getMessage());
        }
    }

    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {
        try {
            String loc = "Prague";
            String appid = "b48908b86c2a129e60b2f84c8a1c8d67";
            host = "api.openweathermap.org";
            resource = "/data/2.5/weather?q=" + loc + "&APPID=" + appid +"&units=metric";
            s = new Socket(host, PORT);

            get();
        } catch (IOException e) {
            System.err.println("Problem: " + e.getMessage());
        }
    }

}
